lib/core/reporters/helpers/process-aggregate.js

```diff
@@ -86,23 +76,31 @@ function normalizeRelatedNodes(node, options) {
       .filter(checkRes => Array.isArray(checkRes.relatedNodes))
       .forEach(checkRes => {
         checkRes.relatedNodes = checkRes.relatedNodes.map(relatedNode => {
-          const res = {
-            html: relatedNode?.source ?? 'Undefined'
-          };
-          if (options.elementRef && !relatedNode?.fromFrame) {
-            res.element = relatedNode?.element ?? null;
-          }
-          if (options.selectors !== false || relatedNode?.fromFrame) {
-            res.target = relatedNode?.selector ?? [':root'];
-          }
-          if (options.ancestry) {
-            res.ancestry = relatedNode?.ancestry ?? [':root'];
-          }
-          if (options.xpath) {
-            res.xpath = relatedNode?.xpath ?? ['/'];
-          }
-          return res;
+          return trimElementSpec(relatedNode, options);
         });
       });
   });
 }
+
+function trimElementSpec(elmSpec = {}, runOptions) {
+  elmSpec = nodeSerializer.dqElmToSpec(elmSpec);
```

Comment:
I think this is going to have the effect of doing the expensive calculations even if `runOptions` disables them, which we probably want to avoid for perf.

The first way I'd probably explore to work around that would be to have `DqElement.toJSON` omit properties that don't need to be calculated based on `runOptions`. I think this would probably need to involve passing runOptions through `nodeSerializer` and `toJSON` as a parameter, so that this final call right before reporting can pass `runOptions` as-is but the `toSpec` calls at frame boundaries can instead pass `{ ...runOptions, ancestry: true }` as to ensure that after methods have the input data they need.

I'm also a little hesitant to leave this out until the point where we're within one of our own reporter implementations. I think it might make more sense to add a `mapRawResults` just before passing raw results to reporters, rather than doing it here within `processAggregate`; otherwise, custom reporters not based on process aggregate will have to know to add the new `toSpec` step themselves.

Reply:
I think it might be better to have `DqElement.setRunOptions` save the run options for `ancestry` and `xpath` and then in the `toJSON` function it doesn't call those getters if the option is not set. That way we don't need to pass in anything when we call `toJSON` and anything that still needs access to those properties still has them.

Reply:
Sorted. I ended up doing this is part of my `dqElmToSpec` call in `processAggregate`. That seemed like the simplest solution. Since that's the only place I want this behavior, I figured that's the placer to do it. Doing this in toJSON, or runRules (before `processAggregate`) would mean it impacts runPartial / collectResultsFromFrame as well, which I'd then have to unset.

I think we'll want to revisit this when we pick up https://github.com/dequelabs/axe-core/issues/4124. It does make sense to move this filtering logic into runRules, but we have to consider things like which checks have after methods to really know what can be skipped and what can't be.